test/xds/xds_server_certificate_providers_test.go

```diff
@@ -0,0 +1,477 @@
+	defer cancel()
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a client that uses insecure creds and verifiy RPCs.
+	cc, err := grpc.Dial(lis.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+
+	client := testgrpc.NewTestServiceClient(cc)
+	if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.WaitForReady(true)); err != nil {
+		t.Fatalf("EmptyCall() failed: %v", err)
+	}
+}
+
+// Tests the case where the bootstrap configuration contains no certificate
+// providers, and xDS credentials with an insecure fallback is specified at
+// server creation time. The management server is configured to return a
+// server-side xDS Listener resource with mTLS security configuration. The xDS
+// client is expected to NACK this resource and therefore he xDS-enabled gRPC
+// server is expected to not enter "serving" mode.
+func (s) TestServerSideXDS_WithNoCertificateProvidersInBootstrap_Failure(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Spin up an xDS management server that pushes on a channel when it
+	// receives a NACK for an LDS response.
+	nackCh := make(chan struct{}, 1)
+	mgmtServer, err := e2e.StartManagementServer(e2e.ManagementServerOptions{
+		OnStreamRequest: func(_ int64, req *v3discoverypb.DiscoveryRequest) error {
+			if req.GetTypeUrl() != "type.googleapis.com/envoy.config.listener.v3.Listener" {
+				return nil
+			}
+			if req.GetErrorDetail() == nil {
+				return nil
+			}
+			select {
+			case nackCh <- struct{}{}:
+			case <-ctx.Done():
+			}
+			return nil
+		},
+		AllowResourceSubset: true,
+	})
+	if err != nil {
+		t.Fatalf("Failed to start management server: %v", err)
+	}
+	defer mgmtServer.Stop()
+
+	// Create bootstrap configuration with no certificate providers.
+	nodeID := uuid.New().String()
+	bs, err := bootstrap.Contents(bootstrap.Options{
+		NodeID:                             nodeID,
+		ServerURI:                          mgmtServer.Address,
+		ServerListenerResourceNameTemplate: e2e.ServerListenerResourceNameTemplate,
+	})
+	if err != nil {
+		t.Fatalf("Failed to create bootstrap configuration: %v", err)
+	}
+
+	// Configure xDS credentials to be used on the server-side.
+	creds, err := xdscreds.NewServerCredentials(xdscreds.ServerOptions{FallbackCreds: insecure.NewCredentials()})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Initialize an xDS-enabled gRPC server and register the stubServer on it.
+	// Pass it a mode change server option that pushes on a channel the mode
+	// changes to "not serving".
+	modeCh := make(chan struct{})
+	modeChangeOpt := xds.ServingModeCallback(func(addr net.Addr, args xds.ServingModeChangeArgs) {
+		if args.Mode == connectivity.ServingModeServing {
+			close(modeCh)
+		}
+	})
```

Comment:
Would rather call modeCh. Also, this sends only once by closing it (and will panic if you close twice). Thinking about it, I made this an event for my server side e2e tests for the server side fix serving := grpcsync.NewEvent()...if args.Mode == connectivity.ServingModeServing {
			serving.Fire()
		}. I think that this semantically makes more sense here (and below).

Reply:
> Would rather call modeCh

It is called `modeCh` already. Don't understand this comment.

> Also, this sends only once by closing it (and will panic if you close twice).

Since, we don't expect this to even happen once, I'm not too worried about a double close.

> I think that this semantically makes more sense here (and below)

The `grpcsync.Event` provides a way to spot check if the event has fired *and* a way to block on the event firing. Since we only need the latter here, I feel a channel is more appropriate and it does not make the code harder to read in anyway. So, I would like to stick to using a channel here, unless you have a very good reason for using a `grpcsync.Event` instead.

Reply:
Sorry servingModeCh. Hmmmm I guess that a channel does make sense here since this is something we expect *not to happen* vs. to happen (i.e. wait till serving until RPC happens)

Reply:
Renamed to `servingModeCh`.